server/src/main/java/io/spine/server/EnvSetting.java

```diff
@@ -0,0 +1,127 @@
+import io.spine.base.EnvironmentType;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A mutable value that may differ between {@linkplain EnvironmentType environment types}.
+ *
+ * <p>For example:
+ * <pre>
+ * {@code
+ * EnvSetting<StorageFactory> storageFactory = new EnvSetting<>();
+ * storageFactory.use(InMemoryStorageFactory.newInstance(), Production.type());
+ *
+ * assertThat(storageFactory.value(Production.type())).isPresent();
+ * assertThat(storageFactory.value(Tests.type())).isEmpty();
+ * }
+ * </pre>
+ *
+ * <p>{@code EnvSetting} values do not determine the environment themselves: it's up to the
+ * caller to ask for the appropriate one.
+ *
+ * @param <P>
+ *         the type of value
+ */
+public final class EnvSetting<P> {
+
+    private final Map<EnvironmentType, P> settingValue = new HashMap<>();
+
+    /**
+     * Returns the value for the specified environment if it was set, an empty {@code Optional}
+     * otherwise.
+     */
+    Optional<P> value(EnvironmentType type) {
+        return Optional.ofNullable(settingValue.get(type));
+    }
+
+    /**
+     * Runs the specified operations against the value of the specified environment if it's
+     * present, does nothing otherwise.
+     *
+     * <p>If you wish to run an operation that doesn't throw, use {@code
+     * value(envType).ifPresent(operation)}.
+     *
+     * @param operation
+     *         operation to run
+     */
+    void ifPresentForEnvironment(EnvironmentType type, ThrowingConsumer<P> operation)
+            throws Exception {
+        P settingValue = this.settingValue.get(type);
+        if (settingValue != null) {
+            operation.accept(settingValue);
+        }
+    }
+
+    /**
+     * If the value for the specified environment is set, just returns it.
+     *
+     * <p>If it is not set, runs the specified supplier, configures and returns the supplied value.
+     */
+    P assignOrDefault(Supplier<P> defaultValue, EnvironmentType type) {
+        checkNotNull(defaultValue);
+        checkNotNull(type);
+        if (settingValue.containsKey(type)) {
+            return settingValue.get(type);
+        } else {
+            P value = defaultValue.get();
+            this.use(value, type);
+            return value;
+        }
+    }
+
+    /** Changes the value for all environments to {@code null}. */
+    void reset() {
+        this.settingValue.clear();
+    }
+
+    /**
+     * Sets the specified value for the specified environment.
+     *
+     * @param value
+     *         value to assign to one of environments
+     */
+    public void use(P value, EnvironmentType type) {
+        checkNotNull(value);
+        checkNotNull(type);
+        this.settingValue.put(type, value);
+    }
+
+    /**
+     * Represents an operation over a value that returns no result and may finish with an error.
```

Comment:
If this is an operation, let's call it so.

Reply:
I disagree. 

`java.util.function.Consumer` Javadoc also says that it's an `operation`. 

Moreover, I think the name immediately tells a developer what this is: it's a regular consumer, except it can `throw`.

Reply:
This name is way too generic for this particular case. That's why I proposed to name it more specifically.